File: trunk/sampleapps/vmailreader/SDK/TokBoxVideo.php
<?php

require_once 'API_Config.php';
require_once 'TokBoxApi.php';
require_once 'TokBoxVMail.php';

class TokBoxVideo {

	public static function getVMailSent($apiObj, $startPage = 0, $count = 10) {
		return self::getFeed($apiObj, "vmailSent", $startPage, $count);
	}
	
	public static function getVmailRecv($apiObj, $startPage = 0, $count = 10) {
		return self::getFeed($apiObj, "vmailRecv", $startPage, $count);
	}
	
	public static function getCallEvent($apiObj, $startPage = 0, $count = 10) {
		return self::getFeed($apiObj, "callEvent", $startPage, $count);
	}
	
	public static function getPublicPost($apiObj, $startPage = 0, $count = 10) {
		return self::getFeed($apiObj, "vmailPostPublic", $startPage, $count);;
	}
	
	public static function getPublicRecv($apiObj, $startPage = 0, $count = 10) {
		return self::getFeed($apiObj, "vmailPostRecv", $startPage, $count);
	}

	public static function generateRecorderEmbedCode($isGuest=false, $width="322", $height="321") {
		$htmlCode = "";
		
		if($isGuest) {
			$apiObj = new TokBoxApi(API_Config::PARTNER_KEY, API_Config::PARTNER_SECRET);
			$apiObj->updateToken($apiObj->getRequestToken(API_Config::CALLBACK_URL));
		
			$htmlCode .= "<script language=\"javascript\" src=\"SDK/js/TokBoxScript.js\"></script>\n";
			$htmlCode .= "<body onclick=\"setToken('".$apiObj->getAuthToken()."');\">\n";			
		}

		$htmlCode .= "<object  width=\"$width\" height=\"$height\">\n".
					"\t<param name=\"movie\" value=\"".API_CONFIG::API_SERVER.TokBoxApi::$API_SERVER_RECORDER_WIDGET."\"></param>\n".
					"\t<param name=\"allowFullScreen\" value=\"true\"></param>\n".
					"\t<param name=\"allowScriptAccess\" value=\"true\"></param>\n".
					"\t<embed id=\"tbx_recorder\" src=\"".API_CONFIG::API_SERVER.TokBoxApi::$API_SERVER_RECORDER_WIDGET."\"\n". 
					"\t\ttype=\"application/x-shockwave-flash\"\n".
					"\t\tallowfullscreen=\"true\"\n".
					"\t\tallowScriptAccess=\"always\"\n".
					"\t\twidth=\"$width\"\n". 
					"\t\theight=\"$height\"\n".
					"\t>\n".
					"\t</embed>\n".
					"</object>\n";
	
		return $htmlCode;
	}
	
	public static function generatePlayerEmbedCode($messageId, $width="425", $height="344", $autoPlay="true") {
		$htmlCode =	"<object  width=\"$width\" height=\"$height\">".
					"\t<param name=\"movie\" value=\"".API_CONFIG::API_SERVER.TokBoxApi::$API_SERVER_PLAYER_WIDGET."\"></param>".
					"\t<param name=\"allowFullScreen\" value=\"true\"></param>".
					"\t<param name=\"allowScriptAccess\" value=\"true\"></param>".
					"\t<param name=\"flashvars\" value=\"targetVmail=$messageId&autoPlay=$autoPlay\"></param>".
					"\t<embed id=\"tbx_player\" src=\"".API_CONFIG::API_SERVER.TokBoxApi::$API_SERVER_PLAYER_WIDGET."\"". 
					"\t\ttype=\"application/x-shockwave-flash\"". 
					"\t\tallowfullscreen=\"true\"".
					"\t\tallowScriptAccess=\"always\"".
					"\t\tflashvars=\"targetVmail=$messageId&autoPlay=$autoPlay\"".
					"\t\twidth=\"$width\"". 
					"\t\theight=\"$height\"".
					"\t>".
					"\t</embed>".
					"</object>";
					
		return $htmlCode;
	}

	private static function getFeed($apiObj, $type="all", $startPage = 0, $count = 10) {
		$feed = $apiObj->getFeed($apiObj->getJabberId(), $type, $startPage, $count);
		
		if(!$feed) {
			throw new Exception("Unable to connect to ".API_Config::API_SERVER.". Please check to make sure API calls are executing properly");
		}
		
		$feedXml = simplexml_load_string($feed);
		
		if($feedXml->error) {
			throw new Exception($feedXml->error, (int)$feedXml->error['code']);			
		}
		
		$itemsArray = array();
		
		foreach($feedXml->feed->item as $item) {
			foreach($item->videoMail as $videoMail) {
				$vmail = new TokBoxVmail($videoMail['vmail_id']);
			
				$vmail->setVmailType($item['type']);
				$vmail->setVmailBatchId($videoMail->content['batch_id']);
				$vmail->setVmailImgUrl($videoMail->image);
				$vmail->setVmailFlvUrl($videoMail->content);
				$vmail->setVmailMessageId($videoMail->content['message_id']);
				
				$recipients = array();
				foreach($videoMail->recipients->recipient as $recipient) {
					$user = new TokBoxVmailUser($recipient['userid'], $recipient, false, $recipient['time_read']);
					$recipients[] = $user;
				}
		
				$vmail->setVmailRecipients($recipients);
	
				$senders = array();
				foreach($videoMail->senders->sender as $sender) {
					$user = new TokBoxVmailUser($sender['userid'], $sender, true);
					$senders[] = $user;
				}
	
				$vmail->setVmailSenders($senders);
				$vmail->setVmailText($videoMail->text);
				$vmail->setVmailTimeRead($videoMail['time_read']);
				$vmail->setVmailTimeSent($videoMail['time_sent']);
	
				$itemsArray[] = $vmail;
			}
		}
		
		return $itemsArray;
	}
}

File: SDKs/Java/src/com/tokbox/api/TokBoxVideo.java
package com.tokbox.api;

import java.io.IOException;
import java.util.ArrayList;
import java.util.logging.Logger;

import javax.xml.parsers.ParserConfigurationException;

import org.w3c.dom.Document;
import org.w3c.dom.NamedNodeMap;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import com.tokbox.exception.TokBoxRuntimeException;
import com.tokbox.object.TokBoxVMail;
import com.tokbox.object.TokBoxVMailUser;
import com.tokbox.util.TokBoxUtils;

public class TokBoxVideo {

	public static final String VMAIL_SENT = "vmailSent";
	public static final String VMAIL_RECV = "vmailRecv";
	public static final String CALL_EVENT = "callEvent";
	public static final String VMAIL_POST_PUBLIC = "vmailPostPublic";
	public static final String VMAIL_POST_RECV = "vmailPostRecv";

	private static Logger logger;

	public static ArrayList<TokBoxVMail> getVMailSent(TokBoxAPI apiObj, int startPage, int count) {
		return TokBoxVideo.getFeed(apiObj, TokBoxVideo.VMAIL_SENT, startPage, count);
	}

	public static ArrayList<TokBoxVMail> getVMailRecv(TokBoxAPI apiObj, int startPage, int count) {
		return TokBoxVideo.getFeed(apiObj, TokBoxVideo.VMAIL_RECV, startPage, count);
	}

	public static ArrayList<TokBoxVMail> getPublicPost(TokBoxAPI apiObj, int startPage, int count) {
		return TokBoxVideo.getFeed(apiObj, TokBoxVideo.VMAIL_POST_PUBLIC, startPage, count);
	}

	public static ArrayList<TokBoxVMail> getPublicRecv(TokBoxAPI apiObj, int startPage, int count) {
		return TokBoxVideo.getFeed(apiObj, TokBoxVideo.VMAIL_POST_RECV, startPage, count);
	}

	private static ArrayList<TokBoxVMail> getFeed(TokBoxAPI apiObj, String type, int startPage, int count) {
		ArrayList<TokBoxVMail> feedItems = null;
		if(API_Config.DEBUG_MODE) {
			TokBoxVideo.logger = Logger.getLogger("com.tokbox.api.TokBoxVideo");
		}

		Document document = null;
		try {		
			if(TokBoxUser.isValid(apiObj, API_Config.PARTNER_SECRET)) {
				document = TokBoxUtils.setupDocument(apiObj.getFeed(apiObj.getJabberId(), type, Integer.toString(startPage), Integer.toString(count), "", "", ""));

				feedItems = new ArrayList<TokBoxVMail>();
				Node feedNode = TokBoxUtils.parseXML("feed", document.getElementsByTagName("feed"));
				NodeList itemNodes = feedNode.getChildNodes();
				int itemNodesLength = itemNodes.getLength();
				for(int i = 0; i < itemNodesLength; i++ ) {
					Node itemNode = itemNodes.item(i);

					if(itemNode.getNodeName().equals("item")) {
						TokBoxVMail newItem = new TokBoxVMail();

						NamedNodeMap itemAttributes = itemNode.getAttributes();
						int itemAttributesLength = itemAttributes.getLength();
						for(int j = 0; j < itemAttributesLength; j++) {
							Node attr = itemAttributes.item(j);
							
							if(attr.getNodeName().equals("type")) {
								newItem.setVmailType(attr.getTextContent());
								break;
							}
						}
						
						Node videomailNode = TokBoxUtils.parseXML("videoMail", itemNode.getChildNodes());
						NamedNodeMap videomailAttributes = videomailNode.getAttributes();
						int videomailAttributesLength = videomailAttributes.getLength();
						for(int j = 0; j < videomailAttributesLength; j++) {
							Node attr = itemAttributes.item(j);
							
							if(attr.getNodeName().equals("vmailId")) {
								newItem.setVmailId(attr.getTextContent());
								break;
							}
						}

						newItem.setVmailTimeRead(TokBoxUtils.parseXML("timeRead", videomailNode.getChildNodes()).getTextContent());
						newItem.setVmailTimeSent(TokBoxUtils.parseXML("timeSent", videomailNode.getChildNodes()).getTextContent());

						Node contentNode = TokBoxUtils.parseXML("content", videomailNode.getChildNodes());
						NamedNodeMap contentAttributes = contentNode.getAttributes();
						int contentAttributesLength = contentAttributes.getLength();
						for(int j = 0; j < contentAttributesLength; j++) {
							Node attr = itemAttributes.item(j);
							
							if(attr.getNodeName().equals("batchId")) {
								newItem.setVmailBatchId(attr.getTextContent());
								break;
							}
						}

						newItem.setVmailText(TokBoxUtils.parseXML("text", contentNode.getChildNodes()).getTextContent());
						newItem.setVmailMessageId(TokBoxUtils.parseXML("messageId", contentNode.getChildNodes()).getTextContent());
						newItem.setVmailFlvUrl(TokBoxUtils.parseXML("video", contentNode.getChildNodes()).getTextContent());
						newItem.setVmailImgUrl(TokBoxUtils.parseXML("image", contentNode.getChildNodes()).getTextContent());

						Node usersNode = TokBoxUtils.parseXML("users", videomailNode.getChildNodes());
						Node sendersNode = TokBoxUtils.parseXML("sender", usersNode.getChildNodes());

						NamedNodeMap sendersAttributes = sendersNode.getAttributes();
						int sendersAttributesLength = sendersAttributes.getLength();
						for(int j = 0; j < sendersAttributesLength; j++) {
							Node attr = sendersAttributes.item(j);

							if(attr.getNodeName().equals("jabberId")) {
								TokBoxVMailUser sender = new TokBoxVMailUser(attr.getTextContent(), TokBoxUtils.parseXML("senderName", sendersNode.getChildNodes()).getTextContent(), true, null);
								
								newItem.setVmailSenders(sender);
								break;
							}
						}

						ArrayList<TokBoxVMailUser> recipients = new ArrayList<TokBoxVMailUser>();

						Node recipientsNode = TokBoxUtils.parseXML("recipients", usersNode.getChildNodes());
						NodeList recipientsList = recipientsNode.getChildNodes();
						int recipientsListCount = recipientsList.getLength();
						for(int j = 0; j < recipientsListCount; j++) {
							Node recipientNode = recipientsList.item(j);
							
							if(recipientNode.getNodeName().equals("recipient")) {
								NamedNodeMap recipientAttributes = recipientNode.getAttributes();
								int recipientAttributesLength = recipientAttributes.getLength();
								for(int k = 0; k < recipientAttributesLength; k++) {
									Node attr = recipientAttributes.item(k);
									
									if(attr.getNodeName().equals("jabberId")) {
										TokBoxVMailUser recipient = new TokBoxVMailUser(attr.getTextContent(), TokBoxUtils.parseXML("recipientName", recipientNode.getChildNodes()).getTextContent(), false, TokBoxUtils.parseXML("timeRead", recipientNode.getChildNodes()).getTextContent());
										recipients.add(recipient);
									}
								}

							}
						}
						
						newItem.setVmailRecipients(recipients);
						
						feedItems.add(newItem);
					}
				}
			}
			else {
				throw new TokBoxRuntimeException("The user is not properly validated");
			}
		} catch(TokBoxRuntimeException e) {
			if(API_Config.DEBUG_MODE) {
				TokBoxVideo.logger.severe(e.getMessage());
				e.printStackTrace();
			}
			throw e;
		} catch (ParserConfigurationException e) {
			if(API_Config.DEBUG_MODE) {
				TokBoxVideo.logger.severe(e.getMessage());
				e.printStackTrace();
			}
			throw new TokBoxRuntimeException("Configuration Error", e);
		} catch (SAXException e) {
			if(API_Config.DEBUG_MODE) {
				TokBoxVideo.logger.severe(e.getMessage());
				e.printStackTrace();
			}
			throw new TokBoxRuntimeException("Error parsing response XML", e);
		} catch (IOException e) {
			if(API_Config.DEBUG_MODE) {
				TokBoxVideo.logger.severe(e.getMessage());
				e.printStackTrace();
			}
			throw new TokBoxRuntimeException("Communications error", e);
		}

		return feedItems;
	}
}